addons/maintainence-agent/maintenance_mode_controller.go

```diff
@@ -0,0 +1,266 @@
+			if scaleDown {
+				logger.Info("Scaling down RBD mirror deployments")
+				replicas = 0
+			} else {
+				logger.Info("Scaling up RBD mirror deployments")
+				replicas = 1
+			}
+			_, err := r.scaleRBDMirrorDeployment(ctx, clusters, replicas)
+			if err != nil {
+				_, statusErr := r.updateStatus(ctx, mmode, mode, ramenv1alpha1.MModeStateError, err)
+				if statusErr != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to update error status %v while having error %v", statusErr, err)
+				}
+				return ctrl.Result{}, err
+			}
+			_, err = r.updateStatus(ctx, mmode, mode, ramenv1alpha1.MModeStateCompleted, nil)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		default:
+			return ctrl.Result{}, fmt.Errorf("no actions found for %q mode", mode)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ModeReconciler) scaleRBDMirrorDeployment(ctx context.Context, clusters []rookv1.CephCluster, replicas int) (ctrl.Result, error) {
+	for _, cephCluster := range clusters {
+		deployments, err := GetDeploymentsStartingWith(ctx, r.SpokeClient, cephCluster.Namespace, RBDMirrorDeploymentNamePrefix)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		for _, deploymentName := range deployments {
+			err = scaleDeployment(ctx, r.SpokeClient, deploymentName, cephCluster.Namespace, int32(replicas))
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ModeReconciler) updateStatus(ctx context.Context, mmode ramenv1alpha1.MaintenanceMode, mode ramenv1alpha1.MMode, state ramenv1alpha1.MModeState, err error) (ctrl.Result, error) {
+	mmode.Status.State = state
+	mmode.Status.ObservedGeneration = mmode.Generation
+	mmode.Status.Conditions = append(mmode.Status.Conditions, newCondition(state, mode, mmode, err))
+	err = r.SpokeClient.Status().Update(ctx, &mmode)
+	return ctrl.Result{}, err
+}
+
+func newCondition(state ramenv1alpha1.MModeState, mode ramenv1alpha1.MMode, mmode ramenv1alpha1.MaintenanceMode, e error) metav1.Condition {
+	var reason string
+	var message string
+
+	switch state {
+	case ramenv1alpha1.MModeStateProgressing:
+		reason = fmt.Sprintf("%s-MaintenanceProgressing", mode)
+		message = "Maintenance of cluster is in progress"
+	case ramenv1alpha1.MModeStateError:
+		reason = fmt.Sprintf("%s-MaintenanceError", mode)
+		message = fmt.Sprintf("%v", e)
+	case ramenv1alpha1.MModeStateCompleted:
+		reason = fmt.Sprintf("%s-MaintenanceCompleted", mode)
+		message = "Maintenance of cluster has completed successfully "
+	default:
+		reason = fmt.Sprintf("%s-MaintenanceUnknown", mode)
+		message = "Maintenance status of cluster is unknown"
+	}
+	return metav1.Condition{
+		Status:             metav1.ConditionStatus(state),
+		ObservedGeneration: mmode.Generation,
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            message,
+	}
+}
+
+func GetDeploymentsStartingWith(ctx context.Context, spokeClient client.Client, namespace string, prefix string) ([]string, error) {
```

Comment:
This is an expensive operation. We should filter by label while listing.

Reply:
added label for the rook mirroring deployment